server/src/file_service.rs

```diff
@@ -305,6 +305,59 @@ pub async fn rename_document(
     }
 }
 
+pub async fn get_document(
+    server_state: &mut ServerState,
+    request: GetDocumentRequest,
+) -> Result<GetDocumentResponse, GetDocumentError> {
+    let transaction = match server_state.index_db_client.transaction().await {
+        Ok(t) => t,
+        Err(e) => {
+            error!("Internal server error! Cannot begin transaction: {:?}", e);
+            return Err(GetDocumentError::InternalError);
+        }
+    };
+
+    let index_result = index_db::get_document_content_status(&transaction, request.id).await;
+    let status = index_result.map_err(|e| match e {
+        index_db::FileError::DoesNotExist => GetDocumentError::DocumentNotFound,
+        _ => {
+            error!(
+                "Internal server error! Cannot get document status from Postgres: {:?}",
+                e
+            );
+            GetDocumentError::InternalError
+        }
+    })?;
+    if status.deleted {
+        return Err(GetDocumentError::DocumentDeleted);
+    }
+    if status.content_version != request.content_version {
+        return Err(GetDocumentError::StaleVersion);
+    }
```

Comment:
deletes in S3 are eventually consistent so we check postgres to see if the document is deleted or out of date - because we never edit objects in S3, these are kind of both just checking if the file is supposed to already be deleted, but distinguishing the cases for potential user-friendliness